### PR TITLE
feat: add Google OAuth2 Authentication for user login

### PR DESCRIPTION
**Pull request by johnson-oragui**
​

This pull request implements Google OAuth2 authentication as a login option. This feature will allow users to log in using their Google accounts, making the authentication process seamless and secure. It enhances user convenience and security, reduces the need for password management, and offers a popular and trusted authentication method.
​

## Description

The changes this makes to the existing code-base is:
 - Created schemas for google oauth response
 - Created a configuration using authlib for google oauth
 - created google_oauth_services module to handle user creation for authenticated users via google oauth
 - create a route to handle the initial call to authenticate with google
 - created a callback-route that handles successful or un-successful response from google authentication server
 - updated main.py to include session management as required for the use of google oauth
 - updated the requirements.txt with google oauth dependencies/modules
 - wrote unittests(pytest) to mock the flow of google oauth (includes teardown)
 - wrote unittests(pytest) to confirm that data are been saved to the database after authentication (includes teardown)

​

## Related Issue
[Google OAuth2 Authentication for User Login](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/33)

​

## Motivation and Context

Google oauth is currently a consistent way users all over different web applications use to access a resource(API or UI) seamlessly. Not having it as a feature to this would make the boilerplate look like it time travelled not into the future, but back into the past. It is now a necessity to have a feature in applications for Oauth to be present. With this feature, a user does not need to go through the hassle of authenticating or accessing a resource by creating new passwords and memorizing it, the user would easily login with an already existing account on google.

​

## How Has This Been Tested?

I Tested the google oauth login route using my browser, I started the application locally, opened the link on my browser, the link redirected me to the login screen of google, I entered my details, it redirected me to the callback route, and my details were saved to the database, and in return I received an access token and a refresh token with user data. The changes  made to the boilerplate did not affect any other part of the code, as it solely depends on just using the create_access_token function already created.

## Screenshots (if appropriate - Postman, etc):

​Opened http://localhost:7001/api/v1/auth/login/google
First screen

Redirected to Google login screen
second screen

Selected my account
third screen

Gave my permission
fourth screen

Google sent response to my callback route, callback handled and generated tokens while returning user data
fifth screen

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
      ​

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Tasks Completed
- [x] Google OAuth2 Login Endpoint: Redirects the user to Google's OAuth2 login page and handles the OAuth2 flow.
- [x] Callback Endpoint: Handles the OAuth2 callback, exchanges authorization code for tokens, and fetches user information.
- [x] Token Management: Issues JWT access and refresh tokens and stores them in relation to the user.
- [x] User Information Update: Fetches and updates user information from Google on re-authentication.
- [x] Error Handling: Properly manages errors related to authentication, token exchange, and user information fetching.

## Expected Outcome
 - The user will be able to easily authenticate with their Google account, receive a token and a refresh token, and use the service with the issued tokens. The system will handle user information updates and error scenarios effectively.

Please review the changes and let me know if there are any adjustments or additional features required. Thank you!

### CODE REVIEW

**Comment by johnson-oragui** [comment]
@CodewithSegNet 

@joboy-dev

**Review by SundayMba — CHANGES_REQUESTED** [review]
please create a different file for google oauth in the route, seperation of concern is very important. 
don't reuse auth.py, use a different file for that.

**Comment by CodewithSegNet** [comment]
Everything looks good. proceed with the changes requested by @sundayMba.

**Review by joboy-dev — CHANGES_REQUESTED** [review]
Create a file in routes for social account login and add your code there

**Comment by johnson-oragui** [comment]
> please create a different file for google oauth in the route, seperation of concern is very important. don't reuse auth.py, use a new file for that.

I have gone ahead and created a new file for google_login and done the necessary changes. Please go ahead and review again.

Thank you.

**Comment by johnson-oragui** [comment]
> Create a file in routes for social account login and add your code there Call it social_auth.py

I have gone ahead and created a new file for google_login and done the necessary changes. Please go ahead and review again.

**Review by joboy-dev — DISMISSED** [review]

**Review by SundayMba — DISMISSED** [review]
okay boss @bolexs

**Review by CodewithSegNet — DISMISSED** [review]

**Review by Goketech — APPROVED** [review]

**Comment by johnson-oragui** [comment]
Amended merge conflicts.

Apologies for going back and forth with this.

Kindly review again @CodewithSegNet  @SundayMba  @Goketech  @joboy-dev

File: api/v1/routes/google_login.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import secrets
+
+from api.db.database import get_db
+from api.core.dependencies.google_oauth_config import google_oauth
+from api.v1.services.google_oauth import GoogleOauthServices
+
+google_auth = APIRouter(prefix="/auth", tags=["Authentication"])
+
+
+@google_auth.get("/login/google")
```
Comment (zxenonx) [inline api/v1/routes/google_login.py:17]: let's use 
```suggestion
@google_auth.get("/google-login")
```

File: api/v1/routes/google_login.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        if not google_response.get("access_token"):
+            raise HTTPException(status_code=400, detail="Authentication Failed")
+
+        # if google has not verified the user email
+        if not google_response.get('userinfo').get('email_verified'):
+            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
+                                detail="Authentication Failed")
+        google_oauth_serivce = GoogleOauthServices()
+        oauth_response = google_oauth_serivce.create(google_response, db)
+        return oauth_response
```
Comment (zxenonx) [inline api/v1/routes/google_login.py:81]:
```suggestion
        return google_oauth_serivce.create(google_response, db)
```

**Review by SundayMba — DISMISSED** [review]
okay boss @bolexs

File: api/v1/services/google_oauth.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+from typing import Annotated, Union
+from api.v1.services.user import user_service
+from api.v1.schemas.google_oauth import Tokens, UserData, StatusResponse
+
+
+class GoogleOauthServices(Service):
+    """
+    Handles database operations for google oauth
+    """
+    def create(self, google_response: dict,
```
Comment (zxenonx) [inline api/v1/services/google_oauth.py:18]: toooo many things in this function - we'll refactor it later.

File: api/v1/services/google_oauth.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    def fetch_all(self, db: Annotated[Session, Depends(get_db)])-> Union[list, HTTPException]:
+        """
+        Retrieves all users information from the oauth table
+
+        Args:
+            db: the database session object for connection
+
+            Returns:
+                list: a list containing all data in oauth table
+        """
```
Comment (zxenonx) [inline api/v1/services/google_oauth.py:126]: Great docstrings.

File: api/v1/services/google_oauth.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            tokens = Tokens(access_token=access_token,
+                            refresh_token=refresh_token,
+                            token_type="bearer")
+
+            success_response = StatusResponse(message="Authentication was successful",
+                                              status="successful",
+                                              statusCode=200,
+                                              tokens=tokens,
+                                              user=user_response)
+            return success_response
```
Comment (zxenonx) [inline api/v1/services/google_oauth.py:189]:
```suggestion
            return StatusResponse(message="Authentication was successful",
                                              status="successful",
                                              statusCode=200,
                                              tokens=tokens,
                                              user=user_response)
```

**Review by zxenonx — CHANGES_REQUESTED** [review]
Thanks @johnson-oragui 
Great job.
I left some comments.

File: api/v1/services/google_oauth.py (hunk 3 above)
Comment (johnson-oragui) [reply api/v1/services/google_oauth.py:18]: Yes Boss.

Thank you

File: api/v1/services/google_oauth.py (hunk 4 above)
Comment (johnson-oragui) [reply api/v1/services/google_oauth.py:126]: Thank you Boss.

**Comment by johnson-oragui** [comment]
> Thanks @johnson-oragui Great job. I left some comments.

Thank you for the review. I have included your suggested refactoring, updated the tests url endpoints too.

Kindly review again.

**Review by zxenonx — APPROVED** [review]
Thanks for the changes @johnson-oragui

**Pull request merged**